File: usage-based-subscriptions/server/php/report_usage.php
<?php
// This code can be run on interval for each active metered subscription
// An example of an interval could be reporting usage once every 24 hours, or even once a minute.

use Stripe\Stripe;
use Ramsey\Uuid\Uuid;
require 'vendor/autoload.php';

$dotenv = Dotenv\Dotenv::create(__DIR__);
$dotenv->load();

// Set your secret key. Remember to switch to your live secret key in production!
// See your keys here: https://dashboard.stripe.com/account/apikeys
Stripe::setApiKey(getenv('STRIPE_SECRET_KEY'));

// Important: your own business logic is needed here before the next step.
// Here is where to pull a record of a customer from your own database.
// Extract the customer's Stripe Subscription Item ID and usage for today from your database record in preparation for reporting to Stripe.
$subscription_item_id = '';
// The usage number you've been keeping track of in your own database for the last 24 hours (or the interval you have set for your needs)
$usage_quantity = 100;

$date = date_create();
$timestamp = date_timestamp_get($date);
// The idempotency key allows you to retry this usage record call if it fails (for example, a network timeout)
$idempotency_key = Uuid::uuid4()->toString();

try {
    \Stripe\SubscriptionItem::createUsageRecord(
        $subscription_item_id,
        [
            'quantity' => $usage_quantity,
            'timestamp' => $timestamp,
        ],
        [
            'idempotency_key' => $idempotency_key,
        ],
    );
} catch (\Stripe\Exception\ApiErrorException $e) {
    echo "usage report failed for item ID $subscription_item_id wuith idempotency key $idempotency_key: $error.toString()";
}

File: usage-based-subscriptions/server/python/report_usage.py
#! /usr/bin/env python3.6

"""
report_usage.py
Stripe Recipe.
This code can be run on an interval (e.g., every 24 hours) for each active
metered subscription.
Python 3.6 or newer required.
"""

import os
import time
import uuid
import stripe
from dotenv import load_dotenv, find_dotenv

# Setup Stripe python client library
load_dotenv(find_dotenv())
stripe.api_key = os.getenv('STRIPE_SECRET_KEY')
#stripe.api_version = os.getenv('STRIPE_API_VERSION')

def report_usage():
    # You need to write some of your own business logic before creating the
    # usage record. Pull a record of a customer from your database
    # and extract the customer's Stripe Subscription Item ID and usage
    # for the day. If you aren't storing subscription item IDs,
    # you can retrieve the subscription and check for subscription items
    # https://stripe.com/docs/api/subscriptions/object#subscription_object-items.
    subscription_item_id = ''
    # The usage number you've been keeping track of in your database for
    # the last 24 hours.
    usage_quantity = 100

    timestamp = int(time.time())
    # The idempotency key allows you to retry this usage record call if it fails.
    idempotency_key = str(uuid.uuid4())

    try:
         stripe.SubscriptionItem.create_usage_record(
            subscription_item_id,
            quantity=usage_quantity,
            timestamp=timestamp,
            action='set',
            idempotency_key=idempotency_key
        )
    except stripe.error.StripeError as e:
        print('Usage report failed for item ID %s with idempotency key %s: %s' % (subscription_item_id, idempotency_key, e))

if __name__ == '__main__':
    report_usage()
